modules/programs/sftpman.nix

```diff
@@ -0,0 +1,101 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.sftpman;
+
+  jsonFormat = pkgs.formats.json { };
+
+  mountOpts = { config, name, ... }: {
+    options = {
+      host = mkOption {
+        type = types.str;
+        description = "The host to connect to.";
+        default = null;
+      };
+      port = mkOption {
+        type = types.port;
+        description = "The port to connect to.";
+        default = 22;
+      };
+      user = mkOption {
+        type = types.str;
+        description = "The username to authenticate with.";
+        default = null;
+      };
+      mountOptions = mkOption {
+        type = types.listOf types.str;
+        description = "Options to pass to sshfs.";
+        default = [ ];
+      };
+      mountPoint = mkOption {
+        type = types.str;
+        description = "The remote path to mount.";
+        default = null;
+      };
+      authType = mkOption {
+        type = types.enum [
+          "password"
+          "publickey"
+          "hostbased"
+          "keyboard-interactive"
+          "gssapi-with-mic"
+        ];
+        description = "The authentication method to use.";
+        default = "publickey";
+      };
+      sshKey = mkOption {
+        type = types.nullOr types.str;
+        description = ''
+          Path to the ssh key to use for authentication.
+          Only applies if authMethod is `publickey`.
+        '';
+        default = cfg.defaultSshKey;
+        apply = sshKey:
+          assert assertMsg (config.authType != "publickey" || sshKey != null)
```

Comment:
It is not recommended to place explicit `assert`s in modules. Use the `assertions` option instead. Something like the following might work (untested):

``` patch
diff --git a/modules/programs/sftpman.nix b/modules/programs/sftpman.nix
index 42fdf8f0..8ecfa7f0 100644
--- a/modules/programs/sftpman.nix
+++ b/modules/programs/sftpman.nix
@@ -52,10 +52,6 @@ let
           Only applies if authMethod is `publickey`.
         '';
         default = cfg.defaultSshKey;
-        apply = sshKey:
-          assert assertMsg (config.authType != "publickey" || sshKey != null)
-            "sftpman ${name}: sshKey must be set when using publickey auth.";
-          sshKey;
       };
       beforeMount = mkOption {
         type = types.str;
@@ -90,6 +86,19 @@ in {
   };
 
   config = mkIf cfg.enable {
+    assertions = [(
+      let
+        hasMissingKey = _: mount: mount.authType == "publickey" && mount.sshKey == null;
+        mountsWithMissingKey = attrNames (filterAttrs hasMissingKey cfg.mounts);
+        mountsWithMissingKeyStr = concatStringsSep ", " mountsWithMissingKey;
+      in {
+        assertion = mountsWithMissingKeyStr == [];
+        message = ''
+          sftpman mounts using authentication type "publickey" but missing 'sshKey': ${mountsWithMissingKeyStr}
+        '';
+      }
+    )];
+
     home.packages = [ cfg.package ];
 
     xdg.configFile = mapAttrs' (name: value:
```

Reply:
With this you can then also add a test of the assertion. See https://github.com/nix-community/home-manager/blob/master/tests/modules/programs/nix-index/assert-on-command-not-found.nix for an example.